File: android/Lib/src/main/kotlin/com/darakeon/dfm/lib/api/entities/Theme.kt
package com.darakeon.dfm.lib.api.entities

import com.darakeon.dfm.lib.R

@Suppress("unused")
enum class Theme(
	val code: Int,
	val style: Int,
) {
	None(0, R.style.Test),
	DarkMagic(1, R.style.DarkMagic),
	DarkSober(2, R.style.DarkSober),
	DarkNature(3, R.style.DarkNature),
	DarkMono(4, R.style.DarkMono),
	LightMagic(-1, R.style.LightMagic),
	LightSober(-2, R.style.LightSober),
	LightNature(-3, R.style.LightNature),
	LightMono(-4, R.style.LightMono);

	companion object {
		fun get(value: Int?) =
			entries.singleOrNull {
				it.code == value
			}
	}
}

File: core/Generic/Theme.cs
// ReSharper disable UnusedMember.Global

using System;

namespace DFM.Generic
{
	public enum Theme
	{
		None = 0,
		DarkMagic = ThemeBrightness.Dark * ThemeColor.Magic,
		LightMagic = ThemeBrightness.Light * ThemeColor.Magic,
		DarkSober = ThemeBrightness.Dark * ThemeColor.Sober,
		LightSober = ThemeBrightness.Light * ThemeColor.Sober,
		DarkNature = ThemeBrightness.Dark * ThemeColor.Nature,
		LightNature = ThemeBrightness.Light * ThemeColor.Nature,
		DarkMono = ThemeBrightness.Dark * ThemeColor.Mono,
		LightMono = ThemeBrightness.Light * ThemeColor.Mono,
	}

	public enum ThemeBrightness
	{
		None = 0,
		Dark = 1,
		Light = -1,
	}

	public enum ThemeColor
	{
		None = 0,
		Magic = 1,
		Sober = 2,
		Nature = 3,
		Mono = 4,
	}

	public static class ThemeX
	{
		public static ThemeBrightness Brightness(this Theme theme)
		{
			return theme > 0 ? ThemeBrightness.Dark :
				theme < 0 ? ThemeBrightness.Light :
				ThemeBrightness.None;
		}

		public static ThemeColor Color(this Theme theme)
		{
			return theme == Theme.None
				? ThemeColor.None
				: (ThemeColor)((Int32)theme / (Int32)theme.Brightness());
		}
	}
}
